async-nats/src/lib.rs

```diff
@@ -769,21 +780,48 @@ pub struct Message {
 pub struct Subscriber {
     _sid: u64,
     receiver: mpsc::Receiver<Message>,
+    sender: mpsc::Sender<ClientOp>,
 }
 
 impl Subscriber {
-    fn new(sid: u64, receiver: mpsc::Receiver<Message>) -> Subscriber {
+    fn new(
+        sid: u64,
+        sender: mpsc::Sender<ClientOp>,
+        receiver: mpsc::Receiver<Message>,
+    ) -> Subscriber {
         Subscriber {
             _sid: sid,
+            sender,
             receiver,
         }
     }
+
+    /// Unsubscribes from subscription, draining all remaining messages.
+    ///
+    /// # Examples
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn unsubscribe() -> Result<(), Box<dyn std::error::Error>> {
+    /// let mut client = async_nats::connect("demo.nats.io").await?;
+    ///
+    /// let mut subscriber = client.subscribe("foo".into()).await?;
+    ///
+    ///  subscriber.unsubscribe();
+    /// # Ok(())
+    /// # }
+    pub fn unsubscribe(self) {
+        drop(self)
+    }
 }
 
 impl Drop for Subscriber {
     fn drop(&mut self) {
         // Can we get away with just closing, and then handling that on the sender side?
         self.receiver.close();
+        let handle = Handle::current();
+        let _entered = handle.enter();
+        futures::executor::block_on(self.sender.send(ClientOp::Unsubscribe { sid: self._sid }))
```

Comment:
I'm almost certain the tokio docs explicitly tell you never to do this, will find the corresponding thing.

Reply:
I hope you're wrong, as then it means we cannot properly unsubscribe on `drop` (unless you have any idea how to).

That means we can do that only on `unsubscribe` and reactively on getting message from dropped sub.

Reply:
I think I am, and was thinking of some pre 1.0 tokio stuff.

Reply:
I think I am, and was thinking of some pre 1.0 tokio stuff.